File: sum_diff.cpp
#include <iostream>
#include <numeric>
using namespace std;

int main()
{
    // int a[] = {2, 4, 2, 3};
    int a[] = {1, 6, 11, 5};
    int n = sizeof(a) / sizeof(a[0]);
    int sum = 0;
    sum = accumulate(a, a + n, sum);
    int ns = sum / 2;
    bool DP[n + 1][ns + 1];

    for (int i = 0; i < n + 1; i++)
    {
        for (int j = 0; j < ns + 1; j++)
        {
            if (j == 0)
                DP[i][j] = true;
            else if (i == 0)
                DP[i][j] = false;
            else if (j < a[i - 1])
                DP[i][j] = DP[i - 1][j];
            else
                DP[i][j] = DP[i - 1][j] || DP[i - 1][j - a[i - 1]];
            cout << DP[i][j] << " ";
        }
        cout << " \n";
    }

    int i = ns;
    while (!DP[n][i])
        i--;
    cout << "The minimum difference possible is: " << sum - 2 * i;
}

File: sum_diff.py
# Problem: Minimum Subset Sum Difference
# Author: Mashu Ajmera
# Approach: Dynamic Programming
# Leetcode: https://leetcode.com/problems/last-stone-weight-ii/
# Github: https://www.geeksforgeeks.org/partition-a-set-into-two-subsets-such-that-the-difference-of-subset-sums-is-minimum/
# YouTube: https://youtu.be/FB0KUhsxXGY

class Solution:
    def lastStoneWeightII(self, stones) -> int:
        n=len(stones)
        s=sum(stones)
        ns=s//2
        dp=[[False for i in range(ns+1)] for j in range(n+1)]

        for i in range(n+1):
            for j in range(ns+1):
                if j==0:
                    dp[i][j]=True
                elif i==0:
                    dp[i][j]=False
                elif j<stones[i-1]:
                    dp[i][j]=dp[i-1][j]
                else:
                    dp[i][j]=dp[i-1][j] or dp[i-1][j-stones[i-1]]
            print(dp[i])
        
        for i in range(ns,-1,-1):
            if dp[n][i]==True:
                return s-2*i

if __name__=="__main__":
    # a=[1,6,11,5] #1
    # a=[3, 1, 4, 2, 2, 1] #1
    # a=[91] #91
    a=[31,26,33,21,40] #5
    s=Solution()
    print(s.lastStoneWeightII(a))
